diff --git a/ltable.c b/ltable.c
index c2d957e9..19674947 100644
--- a/ltable.c
+++ b/ltable.c
@@ -402,7 +402,7 @@ int luaH_next (lua_State *L, Table *t, StkId key) {
 
 static void freehash (lua_State *L, Table *t) {
   if (!isdummy(t)) {
-    size_t bsize = sizenode(t) * sizeof(Node);  /* 'node' size in bytes */
+    size_t bsize = cast_sizet(sizenode(t)) * sizeof(Node);  /* 'node' size in bytes */
     char *arr = cast_charp(t->node);
     if (haslastfree(t)) {
       bsize += sizeof(Limbox);
