diff --git a/testes/table_functions.lua b/testes/table_functions.lua
new file mode 100644
index 00000000..a7f59323
--- /dev/null
+++ b/testes/table_functions.lua
@@ -0,0 +1,17 @@
+t = {
+  function()
+    print('Anonymous table with index key!')
+  end,
+
+  f = function()
+    print('Anonymous table with string key!')
+  end,
+
+  function named_function()
+    print('New named function, success!')
+  end,
+}
+
+t[1]()
+t.f()
+t.named_function()
